files/routes/chat.py

```diff
@@ -1,117 +1,271 @@
+	)
+	if not text: return '', 400
 
-	global messages, total
-	text = data[:1000].strip()
-	if not text: return '', 403
-	text_html = sanitize(text)
+	command = chat_command_regex.match(text)
+	if command:
+		command_name = command.group(1).lower()
+		command_parameters = command.group(2)
+		
+		if command_name in commands:
+			commands[command_name](command_parameters)
+		else:
+			send_system_reply(f"Unknown command: {command_name}")
 
-	data={
-		"avatar": v.profile_url,
-		"username": v.username,
-		"namecolor": v.namecolor,
-		"text": text,
-		"text_html": text_html,
-		"text_censored": text,
-		"time": int(time.time())
-	}
-	
-	if v.shadowbanned:
-		emit('speak', data)
-	else:
-		emit('speak', data, broadcast=True)
-		messages.append(data)
-		messages = messages[-50:]
+		return
 
-	total += 1
+	text_html = sanitize(text)
+	quotes = data['quotes']
+	
+	chat_message = ChatMessage()
+	chat_message.author_id = v.id
+	chat_message.quote_id = quotes
+	chat_message.text = text
+	chat_message.text_html = text_html
+	g.db.add(chat_message)
+	g.db.commit()
 
-	if v.admin_level >= 2:
-		text = text.lower()
-		for i in mute_regex.finditer(text):
-			username = i.group(1)
-			duration = int(int(i.group(2)) * 60 + time.time())
-			muted[username] = duration
+	emit('speak', chat_message.json_speak(), broadcast=True)
 
-	return '', 204
 
 @socketio.on('connect')
-@auth_required
-def connect(v):
+@chat_is_allowed()
+def onConnect(v):
 	if v.username not in online:
 		online.append(v.username)
-		emit("online", online, broadcast=True)
 
+	connected_users.add(request.sid)
+
+	emit('online', get_chat_userlist())
+	emit('catchup', get_chat_messages())
 	emit('typing', typing)
-	return '', 204
+
 
 @socketio.on('disconnect')
-@auth_required
-def disconnect(v):
+@chat_is_allowed()
+def onDisconnect(v):
 	if v.username in online:
 		online.remove(v.username)
-		emit("online", online, broadcast=True)
 
 	if v.username in typing: typing.remove(v.username)
+
+	connected_users.remove(request.sid)
+
 	emit('typing', typing, broadcast=True)
-	return '', 204
+
 
 @socketio.on('typing')
-@auth_required
+@chat_is_allowed()
 def typing_indicator(data, v):
-	if data and v.username not in typing: typing.append(v.username)
-	elif not data and v.username in typing: typing.remove(v.username)
+	if data and v.username not in typing:
+		typing.append(v.username)
+	elif not data and v.username in typing:
+		typing.remove(v.username)
 
 	emit('typing', typing, broadcast=True)
-	return '', 204
 
 
-def close_running_threads():
-	cache.set(f'{SITE}_chat', messages)
-	cache.set(f'{SITE}_total', total)
-	cache.set(f'{SITE}_muted', muted)
-atexit.register(close_running_threads)
+@socketio.on('read')
+@chat_is_allowed()
+def read(data, v):
+	limiter.check()
+	if v.is_banned: return '', 403
+
+	# This value gets truncated at some point in the pipeline and I haven't really spent time to figure out where.
+	# Instead, we just bump it by one.
+	timestamp = datetime.fromtimestamp(int(data) + 1)
```

Comment:
Uh, why trust the client at all?

Reply:
I want to ensure that "messages sent to a non-focused chat window" don't count as "seen", which means I need to get notification that the window has been focused. I don't know *which* window has been focused, or if it fell behind at some point and disconnected, so I just trust the window to send its own data. I can't think of a way that it could overmark messages (besides memory corruption, I guess); if someone wants to hack their client to change their chat markings, okay, sure.